File: macos/Gutenkern/Sources/GutenkernCore/KerningGenerator.swift
import Foundation

public enum Glyph: Equatable, Sendable {
    case character(Character)
    case name(String)
}

public enum PairMode: String, Equatable, Hashable, Sendable, CaseIterable {
    case simple
    case mirror
}

public enum OutputFormat: String, Equatable, Hashable, Sendable, CaseIterable {
    case fontlab
    case glyphs
}

public enum KerningGenerator {
    public static func parse(_ input: String) -> [Glyph] {
        parseTokens(input).map(\.glyph)
    }

    public static func parseTokens(_ input: String) -> [ParsedToken] {
        var tokens: [ParsedToken] = []
        var index = input.startIndex
        var utf16Offset = 0

        while index < input.endIndex {
            let character = input[index]
            if character.isWhitespace {
                utf16Offset += character.utf16.count
                index = input.index(after: index)
                continue
            }

            if character == "/" {
                let nextIndex = input.index(after: index)
                if nextIndex < input.endIndex, input[nextIndex] == "/" {
                    tokens.append(ParsedToken(
                        glyph: .character("/"),
                        start: utf16Offset,
                        length: 2
                    ))
                    utf16Offset += 2
                    index = input.index(after: nextIndex)
                    continue
                }

                let slashOffset = utf16Offset
                utf16Offset += character.utf16.count
                index = nextIndex
                let nameStart = index
                while index < input.endIndex {
                    let next = input[index]
                    if next == "/" || next.isWhitespace {
                        break
                    }
                    utf16Offset += next.utf16.count
                    index = input.index(after: index)
                }
                let name = String(input[nameStart..<index])
                if !name.isEmpty {
                    tokens.append(ParsedToken(
                        glyph: .name(name),
                        start: slashOffset,
                        length: utf16Offset - slashOffset
                    ))
                }
                continue
            }

            let start = utf16Offset
            let length = character.utf16.count
            tokens.append(ParsedToken(
                glyph: .character(character),
                start: start,
                length: length
            ))
            utf16Offset += length
            index = input.index(after: index)
        }

        return tokens
    }

    public static func generate(
        left: String,
        right: String,
        mode: PairMode,
        format: OutputFormat
    ) -> String {
        generate(leftGlyphs: parse(left), rightGlyphs: parse(right), mode: mode, format: format)
    }

    public static func generate(_ input: String, format: OutputFormat) -> String {
        generate(GlyphClassifier.classify(input), format: format)
    }

    public static func generate(_ classified: ClassificationResult, format: OutputFormat) -> String {
        generateSections(classified, format: format)
            .map { $0.joined(separator: "\n\n") }
            .joined(separator: "\n\n\n")
    }

    public static func generateSections(
        _ classified: ClassificationResult,
        format: OutputFormat
    ) -> [[String]] {
        generateRecipeSections(classified, format: format).map(\.groups)
    }

    public static func generateRecipeSections(
        _ classified: ClassificationResult,
        format: OutputFormat
    ) -> [RecipeSection] {
        var sections: [RecipeSection] = []
        for recipe in KerningPlan.recipes {
            guard
                let left = classified.glyphs(for: recipe.left),
                let right = classified.glyphs(for: recipe.right)
            else {
                continue
            }
            let pairGroups = generatePairGroups(
                leftGlyphs: left,
                rightGlyphs: right,
                mode: recipe.mode,
                format: format,
                letterLetterRecipe: KerningScriptFilter.isLetterGroup(recipe.left)
                    && KerningScriptFilter.isLetterGroup(recipe.right)
            )
            if !pairGroups.isEmpty {
                sections.append(RecipeSection(recipe: recipe.line, pairGroups: pairGroups))
            }
        }
        return sections
    }

    public static func generate(
        leftGlyphs: [Glyph],
        rightGlyphs: [Glyph],
        mode: PairMode,
        format: OutputFormat
    ) -> String {
        generateGroups(
            leftGlyphs: leftGlyphs,
            rightGlyphs: rightGlyphs,
            mode: mode,
            format: format
        ).joined(separator: "\n\n")
    }

    public static func generateGroups(
        leftGlyphs: [Glyph],
        rightGlyphs: [Glyph],
        mode: PairMode,
        format: OutputFormat,
        letterLetterRecipe: Bool = false
    ) -> [String] {
        generatePairGroups(
            leftGlyphs: leftGlyphs,
            rightGlyphs: rightGlyphs,
            mode: mode,
            format: format,
            letterLetterRecipe: letterLetterRecipe
        ).map { $0.map(\.display).joined(separator: "\n") }
    }

    public static func generatePairGroups(
        leftGlyphs: [Glyph],
        rightGlyphs: [Glyph],
        mode: PairMode,
        format: OutputFormat,
        letterLetterRecipe: Bool = false
    ) -> [[PairLine]] {
        guard !leftGlyphs.isEmpty, !rightGlyphs.isEmpty else {
            return []
        }

        var groups: [[PairLine]] = []
        groups.reserveCapacity(leftGlyphs.count)

        for leftGlyph in leftGlyphs {
            var groupLines: [PairLine] = []
            groupLines.reserveCapacity(rightGlyphs.count)
            for rightGlyph in rightGlyphs {
                guard KerningScriptFilter.shouldKern(
                    leftGlyph,
                    rightGlyph,
                    letterLetterRecipe: letterLetterRecipe
                ) else {
                    continue
                }
                let sequence: [Glyph]
                switch mode {
                case .simple:
                    sequence = [leftGlyph, rightGlyph]
                case .mirror:
                    sequence = [leftGlyph, rightGlyph, leftGlyph]
                }
                groupLines.append(
                    PairLine(
                        key: formatGlyphs(sequence, as: .fontlab),
                        display: formatGlyphs(sequence, as: format)
                    )
                )
            }
            if !groupLines.isEmpty {
                groups.append(groupLines)
            }
        }

        return groups
    }

    public static func formatGlyphs(_ glyphs: [Glyph], as format: OutputFormat) -> String {
        switch format {
        case .fontlab:
            guard !glyphs.isEmpty else {
                return ""
            }
            return "/" + glyphs.map(plainValue).joined(separator: "/")
        case .glyphs:
            var result = ""
            var previousWasName = false
            for glyph in glyphs {
                switch glyph {
                case .character(let character):
                    if previousWasName {
                        result.append(" ")
                    }
                    result.append(character)
                    previousWasName = false
                case .name(let name):
                    result.append("/")
                    result.append(name)
                    previousWasName = true
                }
            }
            return result
        }
    }

    private static func plainValue(_ glyph: Glyph) -> String {
        switch glyph {
        case .character(let character):
            return String(character)
        case .name(let name):
            return name
        }
    }
}

File: windows/Gutenkern.Core/KerningGenerator.cs
using System.Globalization;

namespace Gutenkern;

public abstract record Glyph;

public sealed record CharacterGlyph(string Value) : Glyph;

public sealed record NameGlyph(string Value) : Glyph;

public enum PairMode
{
    Simple,
    Mirror
}

public enum OutputFormat
{
    FontLab,
    Glyphs
}

public static class KerningGenerator
{
    public static List<Glyph> Parse(string input) =>
        ParseTokens(input).Select(token => token.Glyph).ToList();

    public static List<ParsedToken> ParseTokens(string input)
    {
        var tokens = new List<ParsedToken>();
        var index = 0;

        while (index < input.Length)
        {
            var character = input[index];
            if (char.IsWhiteSpace(character))
            {
                index++;
                continue;
            }

            if (character == '/')
            {
                if (index + 1 < input.Length && input[index + 1] == '/')
                {
                    tokens.Add(new ParsedToken(new CharacterGlyph("/"), index, 2));
                    index += 2;
                    continue;
                }

                var slash = index;
                index++;
                var start = index;
                while (index < input.Length && input[index] != '/' && !char.IsWhiteSpace(input[index]))
                {
                    index++;
                }

                var name = input[start..index];
                if (name.Length > 0)
                {
                    tokens.Add(new ParsedToken(new NameGlyph(name), slash, index - slash));
                }

                continue;
            }

            var elementLength = StringInfo.GetNextTextElementLength(input, index);
            tokens.Add(new ParsedToken(
                new CharacterGlyph(input.Substring(index, elementLength)),
                index,
                elementLength));
            index += elementLength;
        }

        return tokens;
    }

    public static string Generate(string left, string right, PairMode mode, OutputFormat format) =>
        Generate(Parse(left), Parse(right), mode, format);

    public static string Generate(string input, OutputFormat format) =>
        Generate(GlyphClassifier.Classify(input), format);

    public static string Generate(ClassificationResult classified, OutputFormat format) =>
        string.Join("\n\n\n", GenerateSections(classified, format).Select(section => string.Join("\n\n", section)));

    public static List<List<string>> GenerateSections(ClassificationResult classified, OutputFormat format) =>
        GenerateRecipeSections(classified, format).Select(section => section.Groups.ToList()).ToList();

    public static List<RecipeSection> GenerateRecipeSections(ClassificationResult classified, OutputFormat format)
    {
        var sections = new List<RecipeSection>();
        foreach (var recipe in KerningPlan.Recipes)
        {
            if (!classified.TryGet(recipe.Left, out var left) ||
                !classified.TryGet(recipe.Right, out var right))
            {
                continue;
            }

            var pairGroups = GeneratePairGroups(
                left,
                right,
                recipe.Mode,
                format,
                KerningScriptFilter.IsLetterGroup(recipe.Left) && KerningScriptFilter.IsLetterGroup(recipe.Right));
            if (pairGroups.Count > 0)
            {
                sections.Add(new RecipeSection(recipe.Line, pairGroups));
            }
        }

        return sections;
    }

    public static string Generate(
        IReadOnlyList<Glyph> leftGlyphs,
        IReadOnlyList<Glyph> rightGlyphs,
        PairMode mode,
        OutputFormat format) =>
        string.Join("\n\n", GenerateGroups(leftGlyphs, rightGlyphs, mode, format));

    public static List<string> GenerateGroups(
        IReadOnlyList<Glyph> leftGlyphs,
        IReadOnlyList<Glyph> rightGlyphs,
        PairMode mode,
        OutputFormat format,
        bool letterLetterRecipe = false) =>
        GeneratePairGroups(leftGlyphs, rightGlyphs, mode, format, letterLetterRecipe)
            .Select(group => string.Join("\n", group.Select(pair => pair.Display)))
            .ToList();

    public static List<IReadOnlyList<PairLine>> GeneratePairGroups(
        IReadOnlyList<Glyph> leftGlyphs,
        IReadOnlyList<Glyph> rightGlyphs,
        PairMode mode,
        OutputFormat format,
        bool letterLetterRecipe = false)
    {
        if (leftGlyphs.Count == 0 || rightGlyphs.Count == 0)
        {
            return [];
        }

        var groups = new List<IReadOnlyList<PairLine>>(leftGlyphs.Count);
        foreach (var leftGlyph in leftGlyphs)
        {
            var groupLines = new List<PairLine>(rightGlyphs.Count);
            foreach (var rightGlyph in rightGlyphs)
            {
                if (!KerningScriptFilter.ShouldKern(leftGlyph, rightGlyph, letterLetterRecipe))
                {
                    continue;
                }

                var sequence = mode == PairMode.Simple
                    ? new[] { leftGlyph, rightGlyph }
                    : new[] { leftGlyph, rightGlyph, leftGlyph };
                groupLines.Add(new PairLine(
                    FormatGlyphs(sequence, OutputFormat.FontLab),
                    FormatGlyphs(sequence, format)));
            }

            if (groupLines.Count > 0)
            {
                groups.Add(groupLines);
            }
        }

        return groups;
    }

    public static string FormatGlyphs(IReadOnlyList<Glyph> glyphs, OutputFormat format)
    {
        switch (format)
        {
            case OutputFormat.FontLab:
                if (glyphs.Count == 0)
                {
                    return "";
                }

                return "/" + string.Join("/", glyphs.Select(PlainValue));
            case OutputFormat.Glyphs:
                var result = new System.Text.StringBuilder();
                var previousWasName = false;
                foreach (var glyph in glyphs)
                {
                    switch (glyph)
                    {
                        case CharacterGlyph character:
                            if (previousWasName)
                            {
                                result.Append(' ');
                            }

                            result.Append(character.Value);
                            previousWasName = false;
                            break;
                        case NameGlyph name:
                            result.Append('/');
                            result.Append(name.Value);
                            previousWasName = true;
                            break;
                    }
                }

                return result.ToString();
            default:
                throw new ArgumentOutOfRangeException(nameof(format), format, null);
        }
    }

    private static string PlainValue(Glyph glyph) => glyph switch
    {
        CharacterGlyph character => character.Value,
        NameGlyph name => name.Value,
        _ => throw new ArgumentOutOfRangeException(nameof(glyph), glyph, null)
    };
}
